spring-integration-core/src/main/java/org/springframework/integration/config/annotation/BridgeFromAnnotationPostProcessor.java

```diff
@@ -78,4 +79,9 @@ protected MessageHandler createHandler(Object bean, Method method, List<Annotati
 		return handler;
 	}
 
+	@Override
+	protected Object resolveTargetBeanFromMethodWithBeanAnnotation(Method method) {
+		return null;
+	}
```

Comment:
Well, you are missing the fact that these `@BridgeFrom` and `@BridgeTo` are for the `MessageChannel` `@Bean`. There is no any `BridgeHandler` bean definition relationships.

The point here is to not extract bean for the `MessageChannel` since we have the circular dependency issue.

I can guess that we may have similar problem with any other messaging annotation, but that's too late to fix it here.
I have an idea to redo all this logic for `@Bean` methods to the separate `BeanPostProcessor` and don't abuse this one which is applied for the whole `@Configuration` and our attempt to get a bean by method is still too early in the application context phase, although it works in most cases.
Such a big task we can do only in the next version. Right now let's try to keep the state not failing as much as possible.

Does it make sense?

Reply:
Yes, but why not `@BridgeTo` as well?

Reply:
The `@BridgeTo` is slightly different and it really needs an access to the target bean when it creates an endpoint.
I can change the logic there to the `channelResolver` to align API with `BridgeFrom`, but technically it won't make difference because we still are going to get a bean from the application context.

That's why my point is to redo this kind of logic for the messaging annotations on the `@Bean` to some other, much later `BPP` and don't parse these method here in this POJO-specific logic.